### PR TITLE
[SPARK-30819][SPARKR][ML]  Add FMRegressor wrapper to SparkR

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This pull request adds SparkR wrapper for `FMRegressor`:

- Supporting ` org.apache.spark.ml.r.FMRegressorWrapper`.
- `FMRegressionModel` S4 class.
- Corresponding `spark.fmRegressor`, `predict`, `summary` and `write.ml` generics.
- Corresponding docs and tests.

### Why are the changes needed?

Feature parity.

### Does this PR introduce any user-facing change?

No (new API).

### How was this patch tested?

New unit tests.

### CODE REVIEW

**Comment by zero323** [comment]
Jenkins, retest this please.

File: R/pkg/R/mllib_regression.R [hunk 1]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
           function(object, path, overwrite = FALSE) {
             write_internal(object, path, overwrite)
           })
+
+
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: nit: delete extra line?

File: R/pkg/R/mllib_regression.R [hunk 2]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' Factorization Machines Regression Model Model
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: nit: ```Model Model```

File: R/pkg/R/mllib_regression.R [hunk 3]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#'
+#' \code{spark.fmRegressor} fits a factorization regression model against a SparkDataFrame.
+#' Users can call \code{predict} to make
+#' predictions on new data, and \code{write.ml}/\code{read.ml} to save/load fitted models.
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: I guess also mention ```summary``` here?

File: R/pkg/R/mllib_regression.R [hunk 4]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+                                as.integer(maxIter),
+                                as.numeric(stepSize),
+                                as.numeric(tol),
+                                solver,
+                                seed,
+                                stringIndexerOrderType)
+            new("FMRegressionModel", jobj = jobj)
+          })
+
+
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: nit: delete extra line?

File: R/pkg/R/mllib_regression.R [hunk 5]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+            list(
+              coefficients = coefficients,
+              factors = matrix(raw_factors, ncol = factor_size),
+              numFeatures = numFeatures,
+              factorSize = factor_size
+            )
+          })
+
+
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: diddo

File: R/pkg/tests/fulltests/test_mllib_regression.R [hunk 6]
```diff
@@ -551,4 +551,33 @@ test_that("spark.survreg", {
...
+test_that("spark.fmRegressor", {
+  df <- suppressWarnings(createDataFrame(iris))
+
+  model <- spark.fmRegressor(
+    df,  Sepal_Width ~ .,
+    regParam = 0.01, maxIter = 10, fitLinear = TRUE
+  )
+
+  prediction1 <- predict(model, df)
+  expect_is(prediction1, "SparkDataFrame")
```
Comment (huaxingao) [inline R/pkg/tests/fulltests/test_mllib_regression.R]: I guess we may want to check the predict result too?

File: R/pkg/tests/fulltests/test_mllib_regression.R [hunk 7]
```diff
@@ -551,4 +551,33 @@ test_that("spark.survreg", {
...
+
+    prediction2 <- predict(model2, df)
+    expect_equal(
+      collect(prediction1),
+      collect(prediction2)
+    )
+  }
+})
+
+
```
Comment (huaxingao) [inline R/pkg/tests/fulltests/test_mllib_regression.R]: nit: delete extra line

File: examples/src/main/r/ml/fmRegressor.R [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+sparkR.session(appName = "SparkR-ML-fmRegressor-example")
+
+# $example on
+# Load training data
+df <- read.df("data/mllib/sample_linear_regression_data.txt", source = "libsvm")
+training_test <- randomSplit(df, c(0.7, 0.3))
+training <- training_test[[1]]
+test <- training_test[[2]]
+
+
```
Comment (huaxingao) [inline examples/src/main/r/ml/fmRegressor.R]: nit: delete extra line

File: examples/src/main/r/ml/fmRegressor.R [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# To run this example use
+# ./bin/spark-submit examples/src/main/r/ml/decisionTree.R
```
Comment (huaxingao) [inline examples/src/main/r/ml/fmRegressor.R]: change this ```decisionTree.R```?

File: examples/src/main/r/ml/fmRegressor.R [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+training_test <- randomSplit(df, c(0.7, 0.3))
+training <- training_test[[1]]
+test <- training_test[[2]]
+
+
+# Fit a FM regression model
+model <- spark.fmRegressor(training, label ~ features)
+
+# Prediction
+predictions <- predict(model, test)
```
Comment (huaxingao) [inline examples/src/main/r/ml/fmRegressor.R]: same as the classifier example, I guess add ```summary(model)```, ```head(predictions)``` and also add ```sparkR.session.stop()``` in the end?

File: mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      .setFactorSize(factorSize)
+      .setFitLinear(fitLinear)
+      .setRegParam(regParam)
+      .setMiniBatchFraction(miniBatchFraction)
+      .setInitStd(initStd)
+      .setMaxIter(maxIter)
+      .setTol(tol)
+      .setSolver(solver)
+      .setFitIntercept(fitIntercept)
+      .setFeaturesCol(rFormula.getFeaturesCol)
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala]: add ```setStepSize```?

File: mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+    if (seed != null) {
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala]: also check ```seed.length > 0```?

**Comment by zero323** [comment]
Thanks for the reviews @huaxingao!

**Comment by zero323** [comment]
Retest this please.

**Comment by zero323** [comment]
Retest this please.

**Comment by srowen** [comment]
@zero323 I imagine this one is good to go too but now has a merge conflict, b/c FMClassifier was merged

**Comment by zero323** [comment]
Retest this please.

File: R/pkg/R/mllib_regression.R [hunk 13]
```diff
@@ -540,3 +546,147 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' @aliases spark.fmRegressor,SparkDataFrame,formula-method
+#' @name spark.fmRegressor
+#' @seealso \link{read.ml}
+#' @examples
+#' \dontrun{
+#' df <- read.df("data/mllib/sample_linear_regression_data.txt", source = "libsvm")
+#'
+#' # fit Factorization Machines Regression Model
+#' model <- spark.fmRegressor(
+#'            df, label ~ features,
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R:589]: nit: The indents seem not right?

File: mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+import org.apache.spark.ml.feature.RFormula
+import org.apache.spark.ml.r.RWrapperUtils._
+import org.apache.spark.ml.regression.{FMRegressionModel, FMRegressor}
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+private[r] class FMRegressorWrapper private (
+    val pipeline: PipelineModel,
+    val features: Array[String]) extends MLWritable {
+  import FMRegressorWrapper._
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala:36]: nit: remove this?

File: mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    val pipeline = new Pipeline()
+      .setStages(Array(rFormulaModel, fmr))
+      .fit(data)
+
+    new FMRegressorWrapper(pipeline, features)
+  }
+
+  override def read: MLReader[FMRegressorWrapper] = new FMRegressorWrapperReader
+
+  override def load(path: String): FMRegressorWrapper = super.load(path)
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/FMRegressorWrapper.scala:126]: nit: remove this?

**Comment by huaxingao** [comment]
LGTM except a few minor comments.

**Comment by srowen** [comment]
merged to master

**Comment by zero323** [comment]
Thanks @huaxingao @srowen!

**Pull request closed without merging**